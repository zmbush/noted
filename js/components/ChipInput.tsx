// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import * as React from 'react';

import { Autocomplete, TextField } from '@mui/material';

import Tag from './Tag';

interface Props {
  options: string[];
  onChange: (tags: string[]) => void;
  value: string[];
}

const ChipInput = ({ options, value, onChange }: Props) => (
  <div>
    <Autocomplete
      freeSolo
      multiple
      options={options}
      value={value}
      onChange={(_, values) => onChange(values)}
      renderTags={(tags, getTagProps) =>
        tags.map((option, index) => (
          // eslint-disable-next-line react/jsx-props-no-spreading
          <Tag label={option} {...getTagProps({ index })} />
        ))
      }
      renderInput={(params) => (
        <TextField
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...params}
          label='Tags Input'
          InputProps={{ ...params.InputProps, type: 'search' }}
        />
      )}
    />
  </div>
);

export default ChipInput;
