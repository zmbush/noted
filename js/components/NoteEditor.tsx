// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
import colorSyntax from '@toast-ui/editor-plugin-color-syntax';
import '@toast-ui/editor/dist/toastui-editor.css';
import { Editor } from '@toast-ui/react-editor';
import 'codemirror/lib/codemirror.css';
import ChipInput from 'material-ui-chip-input';
import 'tui-color-picker/dist/tui-color-picker.css';

import * as React from 'react';

import SaveIcon from '@mui/icons-material/Save';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardHeader from '@mui/material/CardHeader';
import IconButton from '@mui/material/IconButton';
import Input from '@mui/material/Input';
import { Theme } from '@mui/material/styles';
import { WithStyles } from '@mui/styles';
import createStyles from '@mui/styles/createStyles';
import withStyles from '@mui/styles/withStyles';

import BindKeyboard from 'components/BindKeyboard';
import { NoteData } from 'data/types';

const styles = (theme: Theme) =>
  createStyles({
    editorRoot: {
      [theme.breakpoints.up('sm')]: {
        height: '84vh',
      },
      [theme.breakpoints.down('md')]: {
        height: '100vh',
      },
    },

    titleInput: {
      width: '100%',
    },

    editorContent: {
      position: 'absolute',
      left: 0,
      right: 0,
      bottom: 0,
      top: '50px',
    },
  });

interface Props extends WithStyles<typeof styles> {
  note: NoteData;
  open: boolean;
  onSave: (note: { title: string; body: string; tags: string[] }) => void;
}

type State = {
  title: string;
  body: string;
  tags: string[];
  parent_note_id?: number;
};

class NoteEditor extends React.Component<Props, State> {
  editor: React.RefObject<any>;

  constructor(props: Props) {
    super(props);
    const { note } = this.props;
    const { title, body, tags, parent_note_id: parentNoteId } = note;
    this.state = { title, body, tags, parent_note_id: parentNoteId };
    this.editor = React.createRef();
  }

  componentDidUpdate(oldProps: Props) {
    const { open } = this.props;
    if (open && open !== oldProps.open) {
      const e = this.editor.current;
      if (e) {
        const inst = e.getInstance();
        inst.focus();
        inst.moveCursorToEnd();
      }
    }
  }

  save = (e: React.SyntheticEvent | Event) => {
    e.preventDefault();
    const { onSave } = this.props;
    onSave(this.state);
  };

  addTag = (tag: string) => {
    this.setState((prevState) => ({
      tags: [...prevState.tags, tag],
    }));
  };

  deleteTag = (tag: string, index: number) => {
    const { tags } = this.state;
    tags.splice(index, 1);
    this.setState({
      tags,
    });
  };

  // eslint-disable-next-line react/no-unused-class-component-methods
  hasChanges() {
    const { note } = this.props;
    const current = new Map(Object.entries(this.state));
    const initial = new Map(Object.entries(note));
    return (
      Array.from(current.keys()).filter((key: string) => current.get(key) !== initial.get(key))
        .length > 0
    );
  }

  render() {
    const { classes } = this.props;
    const { title, tags, body } = this.state;
    return (
      <BindKeyboard keys='ctrl+s' callback={this.save}>
        <Card classes={{ root: classes.editorRoot }}>
          <CardHeader
            title={
              <Input
                classes={{ root: classes.titleInput }}
                value={title}
                onChange={(e) => {
                  this.setState({ title: e.target.value });
                }}
                style={{ fontSize: 'inherit' }}
              />
            }
            action={
              <IconButton onClick={this.save} size='large'>
                <SaveIcon />
              </IconButton>
            }
          />
          <CardContent classes={{ root: classes.editorContent }}>
            <ChipInput
              classes={{}}
              placeholder='Tags'
              fullWidth
              dataSource={['arc:Delmirev', 'type:Location', 'type:Character']}
              value={tags}
              onAdd={this.addTag}
              onDelete={this.deleteTag}
            />
            <Editor
              initialValue={body}
              initialEditType='wysiwyg'
              ref={this.editor}
              events={{
                change: () => {
                  if (this.editor.current) {
                    this.setState({
                      body: this.editor.current.getInstance().getMarkdown(),
                    });
                  }
                },
              }}
              height='calc(100% - 40px)'
              usageStatistics={false}
              useCommandShortcut={false}
              plugins={[colorSyntax]}
            />
          </CardContent>
        </Card>
      </BindKeyboard>
    );
  }
}

export const Inner = NoteEditor;
export default withStyles(styles)(NoteEditor);
