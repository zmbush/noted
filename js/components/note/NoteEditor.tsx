// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import * as React from 'react';

import { Save as SaveIcon } from '@mui/icons-material';
import { Card, CardContent, CardHeader, IconButton, Input } from '@mui/material';

import BindKeyboard from 'components/core/BindKeyboard';
import ChipInput from 'components/core/ChipInput';
import MarkdownEditor from 'components/core/markdown/Editor';
import { NoteWithTags, NewNote, UpdateNote } from 'data/types';

type Props = {
  note: NewNote | NoteWithTags;
  onSave: (note: (NewNote | UpdateNote) & Pick<NoteWithTags, 'tags'>) => void;
  onModified: (isModified: boolean) => void;
};

const NoteEditor = ({ note, onSave, onModified }: Props) => {
  const [title, setTitle] = React.useState(note.title);
  const [body, setBody] = React.useState(note.body);
  let noteTags: string[] = [];
  if ('tags' in note) {
    noteTags = note.tags;
  }
  const [tags, setTags] = React.useState(noteTags);

  const notifyChanges = () => {
    const current = new Map(Object.entries({ title, body, tags }));
    const initial = new Map(Object.entries(note));
    onModified(
      Array.from(current.keys()).filter((key: string) => current.get(key) !== initial.get(key))
        .length > 0,
    );
  };

  const save = (e: React.SyntheticEvent | Event) => {
    e.preventDefault();
    onSave({ title, body, tags, parent_note_id: note.parent_note_id });
  };

  return (
    <BindKeyboard keys='ctrl+s' callback={save}>
      <Card
        sx={(theme) => ({
          [theme.breakpoints.up('sm')]: {
            height: '84vh',
          },
          [theme.breakpoints.down('md')]: {
            height: '100vh',
          },
        })}
      >
        <CardHeader
          title={
            <Input
              sx={{ width: '100%' }}
              value={title}
              onChange={(e) => {
                setTitle(e.target.value);
                notifyChanges();
              }}
              style={{ fontSize: 'inherit' }}
            />
          }
          action={
            <IconButton onClick={save} aria-label='Save Note' size='large' tabIndex={-1}>
              <SaveIcon />
            </IconButton>
          }
        />
        <CardContent
          sx={{
            position: 'absolute',
            left: 0,
            right: 0,
            bottom: 0,
            top: '50px',
          }}
        >
          <ChipInput
            value={tags}
            onChange={setTags}
            options={['type:Location', 'type:Character']}
          />
          <MarkdownEditor height='calc(100% - 40px)' body={body} onChange={(v) => setBody(v())} />
        </CardContent>
      </Card>
    </BindKeyboard>
  );
};

export default NoteEditor;
