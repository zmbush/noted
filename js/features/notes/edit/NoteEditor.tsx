// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { createSelector } from '@reduxjs/toolkit';

import * as React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { Save as SaveIcon } from '@mui/icons-material';
import {
  Card,
  CardContent,
  CardHeader,
  IconButton,
  Input,
  Dialog,
  useMediaQuery,
  Theme,
} from '@mui/material';

import { AppState } from 'features/redux/store';

import BindKeyboard from 'components/BindKeyboard';
import ChipInput from 'components/ChipInput';
import ConfirmationDialog from 'components/ConfirmationDialog';
import Loading from 'components/Loading';
import { NoteWithTags, NewNotePayload } from 'data/types';

import { createNote, updateNote } from '../list/api';
import { getSingleNote } from '../list/selectors';

import { NoteEditState, prefix, setEditingNote } from './slice';

const MarkdownEditor = React.lazy(() => import('features/markdown/Editor'));

type NewNoteOrEdit = { id: 'new'; note: NewNotePayload } | { id: number; note: NoteWithTags };

const getNoteEdit = (state: AppState): NoteEditState => state[prefix];
const getEditingNote = createSelector(getNoteEdit, (noteEdit) => noteEdit.editingNote);
const getNoteBeingEdited = createSelector(
  (s: AppState) => s,
  getEditingNote,
  (state, noteEdit): NewNoteOrEdit | null => {
    if (!noteEdit) {
      return null;
    }
    if (noteEdit.id === 'new') {
      const { id, title = '', parent_note_id: parentNoteId } = noteEdit;
      const note: NewNotePayload = { title, body: '', parent_note_id: parentNoteId };
      return { id, note };
    }
    const { id } = noteEdit;
    return { id, note: getSingleNote(state, { noteId: id }) };
  },
);

const NoteEditorInner = ({ id, note }: NewNoteOrEdit) => {
  const [title, setTitle] = React.useState(note.title);
  const [body, setBody] = React.useState(note.body);
  let noteTags: string[] = [];
  if ('tags' in note) {
    noteTags = note.tags;
  }
  const [tags, setTags] = React.useState(noteTags);
  const [confirmCancelEditOpen, setConfirmCancelEditOpen] = React.useState(false);
  const editorFullscreen = useMediaQuery<Theme>((theme) => theme.breakpoints.down('sm'));
  const dispatch = useDispatch();

  const current = new Map(Object.entries({ title, body, tags }));
  const initial = new Map(Object.entries(note));
  const noteIsModified =
    Array.from(current.keys()).filter((key: string) => current.get(key) !== initial.get(key))
      .length > 0;

  const stopEditing = () => dispatch(setEditingNote(null));
  const save = async (e: React.SyntheticEvent | Event) => {
    e.preventDefault();
    stopEditing();

    if (id === 'new') {
      await dispatch(createNote({ title, body, tags, parent_note_id: note.parent_note_id }));
    } else {
      await dispatch(updateNote({ id, title, body, tags }));
    }
  };

  const tryCancelEdit = () => {
    if (noteIsModified) {
      setConfirmCancelEditOpen(true);
    } else {
      stopEditing();
    }
  };

  return (
    <>
      <ConfirmationDialog
        open={confirmCancelEditOpen}
        title='If you close this editor, you will lose your changes.'
        onPositive={stopEditing}
        onNegative={() => setConfirmCancelEditOpen(false)}
      />
      <Dialog
        open
        fullWidth
        maxWidth='lg'
        fullScreen={editorFullscreen}
        onClose={tryCancelEdit}
        data-testid='edit-note-dialog'
      >
        <BindKeyboard keys='ctrl+s' callback={save}>
          <Card
            data-testid='NoteEditor'
            sx={(theme) => ({
              [theme.breakpoints.up('sm')]: {
                height: '84vh',
              },
              [theme.breakpoints.down('md')]: {
                height: '100vh',
              },
            })}
          >
            <CardHeader
              title={
                <Input
                  sx={{ width: '100%' }}
                  value={title}
                  onChange={(e) => setTitle(e.target.value)}
                  style={{ fontSize: 'inherit' }}
                />
              }
              action={
                <IconButton onClick={save} aria-label='Save Note' size='large' tabIndex={-1}>
                  <SaveIcon />
                </IconButton>
              }
            />
            <CardContent
              sx={{
                position: 'absolute',
                left: 0,
                right: 0,
                bottom: 0,
                top: '50px',
              }}
            >
              <ChipInput
                value={tags}
                onChange={setTags}
                options={['type:Location', 'type:Character']}
              />
              <React.Suspense fallback={<Loading />}>
                <MarkdownEditor
                  height='calc(100% - 50px)'
                  body={body}
                  onChange={(v) => setBody(v())}
                />
              </React.Suspense>
            </CardContent>
          </Card>
        </BindKeyboard>
      </Dialog>
    </>
  );
};

const NoteEditor = () => {
  const noteAndId = useSelector(getNoteBeingEdited);
  if (noteAndId === null) {
    return null;
  }
  const { id } = noteAndId;
  if (id === 'new') {
    const { note } = noteAndId;
    return <NoteEditorInner id='new' note={note} />;
  }
  const { note } = noteAndId;
  return <NoteEditorInner id={id} note={note} />;
};

export default NoteEditor;
