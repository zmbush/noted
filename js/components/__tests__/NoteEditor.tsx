// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import * as React from 'react';
import { shallow } from 'enzyme';

import NoteEditor, { Inner } from '../NoteEditor';
import ChipInput from 'material-ui-chip-input';
import IconButton from '@material-ui/core/IconButton';

describe('<NoteEditor />', () => {
  test('matches snapshot', () => {
    let wrapper = shallow(
      <NoteEditor
        open={false}
        note={{
          id: 1,
          title: 'note title',
          body: 'note body',
          tags: ['tag1'],
          created_at: '',
          updated_at: '',
          user_id: 2,
        }}
        onSave={() => {}}
      />
    ).dive();

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({
      note: {
        title: 'new title',
        body: 'new body',
        tags: ['tag2'],
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  test('tag input works', () => {
    let wrapper = shallow(
      <NoteEditor
        open={false}
        note={{
          id: 1,
          title: 'note title',
          body: 'note body',
          tags: ['tag1'],
          created_at: '',
          updated_at: '',
          user_id: 2,
        }}
        onSave={() => {}}
      />
    ).dive();

    const input = wrapper.find(ChipInput);

    input.props().onAdd('test');
    expect((wrapper.state() as any).tags).toContain('test');
    input.props().onDelete('test', 1);
    expect((wrapper.state() as any).tags).not.toContain('test');
  });

  test('switching to open works', () => {
    let wrapper = shallow(
      <NoteEditor
        open={false}
        note={{
          id: 1,
          title: 'note title',
          body: 'note body',
          tags: ['tag1'],
          created_at: '',
          updated_at: '',
          user_id: 2,
        }}
        onSave={() => {}}
      />
    ).dive();
    wrapper.setProps({ open: true });
  });

  test('submitting works', () => {
    let noteSaved;
    let wrapper = shallow(
      <NoteEditor
        open={false}
        note={{
          id: 1,
          title: 'note title',
          body: 'note body',
          tags: ['tag1'],
          created_at: '',
          updated_at: '',
          user_id: 2,
        }}
        onSave={note => {
          noteSaved = note;
        }}
      />
    ).dive();

    wrapper
      .find('WithStyles(CardHeader)')
      .dive()
      .dive()
      .find(IconButton)
      .simulate('click', { preventDefault() {} });

    expect(noteSaved).toEqual({
      title: 'note title',
      body: 'note body',
      tags: ['tag1'],
    });
  });
});
