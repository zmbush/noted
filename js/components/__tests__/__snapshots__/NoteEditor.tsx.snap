// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteEditor /> matches snapshot 1`] = `
<BindKeyboard
  callback={[Function]}
  keys="ctrl+s"
>
  <WithStyles(ForwardRef(Card))
    classes={
      Object {
        "root": "",
      }
    }
  >
    <WithStyles(ForwardRef(CardHeader))
      action={
        <WithStyles(ForwardRef(IconButton))
          onClick={[Function]}
        >
          <Memo />
        </WithStyles(ForwardRef(IconButton))>
      }
      title={
        <WithStyles(ForwardRef(Input))
          classes={
            Object {
              "root": "",
            }
          }
          onChange={[Function]}
          style={
            Object {
              "fontSize": "inherit",
            }
          }
          value="note title"
        />
      }
    />
    <WithStyles(ForwardRef(CardContent))
      classes={
        Object {
          "root": "",
        }
      }
    >
      <WithStyles(ChipInput)
        classes={Object {}}
        dataSource={
          Array [
            "arc:Delmirev",
            "type:Location",
            "type:Character",
          ]
        }
        fullWidth={true}
        onAdd={[Function]}
        onDelete={[Function]}
        placeholder="Tags"
        value={
          Array [
            "tag1",
          ]
        }
      />
      <t
        exts={
          Array [
            "colorSyntax",
          ]
        }
        height="calc(100% - 40px)"
        initialEditType="wysiwyg"
        initialValue="note body"
        onChange={[Function]}
        usageStatistics={false}
        useCommandShortcut={false}
      />
    </WithStyles(ForwardRef(CardContent))>
  </WithStyles(ForwardRef(Card))>
</BindKeyboard>
`;

exports[`<NoteEditor /> matches snapshot 2`] = `
<BindKeyboard
  callback={[Function]}
  keys="ctrl+s"
>
  <WithStyles(ForwardRef(Card))
    classes={
      Object {
        "root": "",
      }
    }
  >
    <WithStyles(ForwardRef(CardHeader))
      action={
        <WithStyles(ForwardRef(IconButton))
          onClick={[Function]}
        >
          <Memo />
        </WithStyles(ForwardRef(IconButton))>
      }
      title={
        <WithStyles(ForwardRef(Input))
          classes={
            Object {
              "root": "",
            }
          }
          onChange={[Function]}
          style={
            Object {
              "fontSize": "inherit",
            }
          }
          value="note title"
        />
      }
    />
    <WithStyles(ForwardRef(CardContent))
      classes={
        Object {
          "root": "",
        }
      }
    >
      <WithStyles(ChipInput)
        classes={Object {}}
        dataSource={
          Array [
            "arc:Delmirev",
            "type:Location",
            "type:Character",
          ]
        }
        fullWidth={true}
        onAdd={[Function]}
        onDelete={[Function]}
        placeholder="Tags"
        value={
          Array [
            "tag1",
          ]
        }
      />
      <t
        exts={
          Array [
            "colorSyntax",
          ]
        }
        height="calc(100% - 40px)"
        initialEditType="wysiwyg"
        initialValue="note body"
        onChange={[Function]}
        usageStatistics={false}
        useCommandShortcut={false}
      />
    </WithStyles(ForwardRef(CardContent))>
  </WithStyles(ForwardRef(Card))>
</BindKeyboard>
`;
